### ISSUE TITLE
Environment variables must override .env settings

### ISSUE THREAD

**Issue by osantana** [henriquebastos/python-decouple#9]
In the Unix world settings and application parameters are usually evaluated in the following order:

1. CLI arguments (eg. `python -u`)
2. Environment variables (eg. `PYTHONUNBUFFERED=1`)
3. Local settings files (eg. `./project/setup.cfg`)
4. User settings files (eg. `$HOME/.setup.cfg`)
5. System settings files (eg. `/etc/setup.cfg`)

But `python-decouple` change this order and use the `.env` file settings even if we've a envvar with diferent value:

https://github.com/henriquebastos/python-decouple/blob/master/decouple.py#L127-L131

I would like to propose the following (backward incompatible) change for the method above:

```python
    def get(self, key):
        try:
            return os.environ[key]
        except KeyError:
            return self.data[key]
```

**Comment by bertonha** [comment]
:+1: This will solve my problems with docker and local run

**Comment by henriquebastos** [comment]
Osvaldo, interesting. This may open decouple to more possibilities.

Christofer, can you detail the problems you're having so I could reproduce
it here?

**Comment by bertonha** [comment]
@henriquebastos in my local machine I use sqlite to develop and on docker use mysql.
in my .env was DB=sqlite.
When I start docker the file `.env` "override" the docker environment variables.

**Comment by henriquebastos** [comment]
Christofer, idealy you should use a different/specific `.env` file for
every instance, and define defaults (*fallbacks*) for some `options` on
your `settings.py`.

`.env` is not the *project* config file (that's the `settings.py`). The
`.env` is the **instance** config file.

**Comment by osantana** [comment]
One problem I have is that I use `.env` with my local settings during development (eg. `DATABASE_URL = mysql://...@localhost /...`) and I also use Docker to run integration tests on my machine. These tests need specific settings provided by envvars (eg. `DATABASE_URL = mysql://...@dbserver/...).

The problem happens when I run the tests in Docker and the .env file appears inside the container. In this case `python-decouple` use this file instead of envvars provided.

**Comment by henriquebastos** [comment]
Got it. it make sense to me that `decouple` supports the *unix style* of overriding `options`.

Should this also happen with `.ini` files?

@osantana can you update a PR with tests?

**Comment by osantana** [comment]
Yes, I can do it.

But this change is backward incompatible, so, it could break something. Is it ok for you?

**Comment by henriquebastos** [comment]
Yes. I understand we're **changin the priorities** on how *decouple*
evaluates an `option`. Today, for `.env` we use `environ` as a *fallback*.
This change will use `environ` to **override** `.env`, aligning `decouple`
with the common unix behavior.

Right now (without actual code), I don't see reasons why we shouldn't do
it. And I'm +1 for that.

About the PR:

1. Make sure you adapt/extend the current tests for this new behavior. It
will be easier to review the patch.

2. Note that *decouple* should still raise the *undefined exception* when
(1) an `option` does not exists (either on file and environ) and (2) no
default parameter was passed to `config`.

Thank you for doing this. :)

**Comment by henriquebastos** [comment]
Just released version 3.0 supporting environment variables overriding config files. This closes #10 and #17.

I've experimented with the `RapositoryChain` patch and with inheritance. In the end it felt like yak shaving and I simply made `RepositoryEnv` and `RepositoryIni` explicitly check on `os.environ`.

From this issue we're still pending the path search. I've thought about it and I would be satisfied with an inversion of dependency of the *search path* on `AutoConfig` and `Config`.

So by default we could search from the `settings.py` path, but users could instantiate their own `AutoConfig` or `Config` with a specific path, like: `config = AutoConfig('my/custom/search/path/')`

However, since I've never needed this I wasn't sure this was an actual useful feature.
